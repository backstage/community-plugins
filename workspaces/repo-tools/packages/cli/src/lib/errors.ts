import chalk from 'chalk';

export class CustomError extends Error {
  get name(): string {
    return this.constructor.name;
  }
}

export class ExitCodeError extends CustomError {
  readonly code: number;

  constructor(code: number, command?: string) {
    super(
      command
        ? `Command '${command}' exited with code ${code}`
        : `Child exited with code ${code}`,
    );
    this.code = code;
  }
}

export function exitWithError(error: Error): never {
  if (error instanceof ExitCodeError) {
    process.stderr.write(`\n${chalk.red(error.message)}\n\n`);
    process.exit(error.code);
  } else {
    process.stderr.write(`\n${chalk.red(`${error.stack}`)}\n\n`);
    process.exit(1);
  }
}
