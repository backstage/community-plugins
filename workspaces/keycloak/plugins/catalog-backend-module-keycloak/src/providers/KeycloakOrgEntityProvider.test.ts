/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type {
  LoggerService,
  SchedulerServiceTaskInvocationDefinition,
  SchedulerServiceTaskRunner,
  SchedulerServiceTaskScheduleDefinition,
} from '@backstage/backend-plugin-api';
import { mockServices, ServiceMock } from '@backstage/backend-test-utils';
import { ErrorLike } from '@backstage/errors';
import type { EntityProviderConnection } from '@backstage/plugin-catalog-node';
import type { JsonObject } from '@backstage/types';

// @ts-ignore
import inclusion from 'inclusion';

import {
  assertLogMustNotInclude,
  authMock,
  CONFIG,
  KeycloakAdminClientMockServerv18,
  KeycloakAdminClientMockServerv24,
  PASSWORD_CONFIG,
} from '../../__fixtures__/helpers';
import { KeycloakOrgEntityProvider } from './KeycloakOrgEntityProvider';

jest.mock('inclusion', () => jest.fn());

const connection = {
  applyMutation: jest.fn(),
  refresh: jest.fn(),
} as unknown as EntityProviderConnection;

class SchedulerServiceTaskRunnerMock implements SchedulerServiceTaskRunner {
  private tasks: SchedulerServiceTaskInvocationDefinition[] = [];
  async run(task: SchedulerServiceTaskInvocationDefinition) {
    this.tasks.push(task);
  }
  async runAll() {
    const abortSignal = jest.fn() as unknown as AbortSignal;
    for await (const task of this.tasks) {
      await task.fn(abortSignal);
    }
  }
}

const scheduler = mockServices.scheduler.mock({
  createScheduledTaskRunner() {
    return new SchedulerServiceTaskRunnerMock();
  },
});

describe.each([
  ['v24', KeycloakAdminClientMockServerv24],
  ['v18', KeycloakAdminClientMockServerv18],
])('KeycloakOrgEntityProvider with %s', (_version, mockImplementation) => {
  let logger: ServiceMock<LoggerService>;
  let keycloakLogger: ServiceMock<LoggerService>;
  let schedule: SchedulerServiceTaskRunnerMock;

  beforeEach(() => {
    jest.clearAllMocks();
    authMock.mockReset();
    keycloakLogger = mockServices.logger.mock();
    logger = mockServices.logger.mock({
      child: () => keycloakLogger,
    });
    inclusion.mockImplementation(() => ({ default: mockImplementation })); // Return the correct mock based on the version
    schedule = scheduler.createScheduledTaskRunner(
      '' as unknown as SchedulerServiceTaskScheduleDefinition,
    ) as SchedulerServiceTaskRunnerMock;
  });

  afterEach(() => {
    for (const log of [logger, keycloakLogger]) {
      assertLogMustNotInclude(log, ['myclientsecret', 'mypassword']); // NOSONAR
    }
  });

  const createProvider = (configData: JsonObject) =>
    KeycloakOrgEntityProvider.fromConfig(
      {
        config: mockServices.rootConfig({ data: configData }),
        logger,
      },
      {
        schedule,
      },
    );

  const runProvider = async (configData: JsonObject) => {
    const keycloak = createProvider(configData);
    for await (const k of keycloak) {
      await k.connect(connection);
      await schedule.runAll();
    }
  };

  it('should mock inclusion', async () => {
    const KeyCloakAdminClient = await inclusion(
      '@keycloak/keycloak-admin-client',
    );
    expect(KeyCloakAdminClient).toEqual({ default: mockImplementation });
  });

  it('should connect', async () => {
    const keycloak = createProvider(CONFIG);
    const result = await Promise.all(
      keycloak.map(async k => await k.connect(connection)),
    );
    expect(result).toEqual([undefined]);
  });

  it('should not read without a connection', async () => {
    const keycloak = createProvider(CONFIG);

    for await (const k of keycloak) {
      await expect(() => k.read()).rejects.toThrow('Not initialized');
    }
    expect(authMock).toHaveBeenCalledTimes(0);
  });

  it('should fail with grantType client_credential, but without client secret', async () => {
    const invalidConfig = {
      catalog: {
        providers: {
          keycloakOrg: {
            default: {
              baseUrl: 'http://localhost:8080',
              clientId: 'myclientid',
            },
          },
        },
      },
    };

    expect(() => createProvider(invalidConfig)).toThrow(
      'clientSecret must be provided when clientId is defined.',
    );
    expect(authMock).toHaveBeenCalledTimes(0);
  });

  it('should read with grantType client_credential', async () => {
    const validConfig = {
      catalog: {
        providers: {
          keycloakOrg: {
            default: {
              baseUrl: 'http://localhost:8080',
              clientId: 'myclientid',
              clientSecret: 'myclientsecret',
            },
          },
        },
      },
    };

    await runProvider(validConfig);

    expect(authMock).toHaveBeenCalledTimes(1);
    expect(authMock).toHaveBeenCalledWith({
      grantType: 'client_credentials',
      clientId: 'myclientid',
      clientSecret: 'myclientsecret',
    });
    expect(connection.applyMutation).toHaveBeenCalledTimes(1);
    expect(
      (connection.applyMutation as jest.Mock).mock.calls,
    ).toMatchSnapshot();
  });

  it('should fail read with grantType username, but without password', async () => {
    const invalidConfig = {
      catalog: {
        providers: {
          keycloakOrg: {
            default: {
              baseUrl: 'http://localhost:8080',
              username: 'myusername',
            },
          },
        },
      },
    };

    expect(() => createProvider(invalidConfig)).toThrow(
      'password must be provided when username is defined.',
    );
    expect(authMock).toHaveBeenCalledTimes(0);
  });

  it('should read with grantType password', async () => {
    await runProvider(PASSWORD_CONFIG);

    expect(authMock).toHaveBeenCalledTimes(1);
    expect(authMock).toHaveBeenCalledWith({
      grantType: 'password',
      clientId: 'admin-cli',
      username: 'myusername',
      password: 'mypassword', // NOSONAR
    });
    expect(connection.applyMutation).toHaveBeenCalledTimes(1);
    expect(
      (connection.applyMutation as jest.Mock).mock.calls,
    ).toMatchSnapshot();
  });

  it('should log a proper error when network connection was refused', async () => {
    // Create an error that contains sensitive information.
    // The afterEach call ensure that this information aren't logged.
    const error = new Error('connect ECONNREFUSED ::1:8080') as ErrorLike;
    error.code = 'ECONNREFUSED';
    error.config = {
      data: 'username=myusername&password=mypassword', // NOSONAR
    };
    error.status = null;
    authMock.mockRejectedValue(error);

    await runProvider(PASSWORD_CONFIG);

    expect(authMock).toHaveBeenCalledTimes(1);
    expect(authMock).toHaveBeenCalledWith({
      grantType: 'password',
      clientId: 'admin-cli',
      username: 'myusername',
      password: 'mypassword', // NOSONAR
    });
    expect(connection.applyMutation).toHaveBeenCalledTimes(0);

    expect(logger.child).toHaveBeenCalledTimes(1);
    expect(keycloakLogger.info).toHaveBeenCalledWith(
      'Reading Keycloak users and groups',
    );
    expect(keycloakLogger.error).toHaveBeenCalledWith(
      'Error while syncing Keycloak users and groups',
      {
        name: 'Error',
        message: 'connect ECONNREFUSED ::1:8080',
        stack: expect.any(String),
      },
    );
  });

  it('should log a proper error when network connection was forbidden', async () => {
    // Create an error that contains sensitive information.
    // The afterEach call ensure that this information aren't logged.
    const error = new Error('Request failed with status code 401') as ErrorLike;
    error.config = {
      data: 'username=myusername&password=mypassword', // NOSONAR
    };
    error.status = 401;
    authMock.mockRejectedValue(error);

    await runProvider(PASSWORD_CONFIG);

    expect(authMock).toHaveBeenCalledTimes(1);
    expect(authMock).toHaveBeenCalledWith({
      grantType: 'password',
      clientId: 'admin-cli',
      username: 'myusername',
      password: 'mypassword', // NOSONAR
    });
    expect(connection.applyMutation).toHaveBeenCalledTimes(0);

    expect(logger.child).toHaveBeenCalledTimes(1);
    expect(keycloakLogger.info).toHaveBeenCalledWith(
      'Reading Keycloak users and groups',
    );
    expect(keycloakLogger.error).toHaveBeenCalledWith(
      'Error while syncing Keycloak users and groups',
      {
        name: 'Error',
        message: 'Request failed with status code 401',
        stack: expect.any(String),
      },
    );
  });
});
