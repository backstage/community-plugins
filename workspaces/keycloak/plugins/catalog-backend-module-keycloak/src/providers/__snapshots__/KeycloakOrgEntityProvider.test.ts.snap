// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeycloakOrgEntityProvider with v18 should read with grantType client_credential 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/id": "59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/realm": "master",
              },
              "name": "jamesdoe",
            },
            "spec": {
              "memberOf": [
                "biggroup",
              ],
              "profile": {
                "email": "jamesdoe@gmail.com",
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/id": "c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/realm": "master",
              },
              "name": "joedoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/id": "2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/realm": "master",
              },
              "name": "johndoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "displayName": "John Doe",
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/id": "9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/realm": "master",
              },
              "name": "biggroup",
            },
            "spec": {
              "children": [
                "subgroup",
              ],
              "members": [
                "jamesdoe",
              ],
              "parent": undefined,
              "profile": {
                "displayName": "biggroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/id": "eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/realm": "master",
              },
              "name": "subgroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": "biggroup",
              "profile": {
                "displayName": "subgroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/id": "557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/realm": "master",
              },
              "name": "emptygroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": undefined,
              "profile": {
                "displayName": "emptygroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
      ],
      "type": "full",
    },
  ],
]
`;

exports[`KeycloakOrgEntityProvider with v18 should read with grantType password 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/id": "59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/realm": "master",
              },
              "name": "jamesdoe",
            },
            "spec": {
              "memberOf": [
                "biggroup",
              ],
              "profile": {
                "email": "jamesdoe@gmail.com",
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/id": "c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/realm": "master",
              },
              "name": "joedoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/id": "2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/realm": "master",
              },
              "name": "johndoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "displayName": "John Doe",
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/id": "9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/realm": "master",
              },
              "name": "biggroup",
            },
            "spec": {
              "children": [
                "subgroup",
              ],
              "members": [
                "jamesdoe",
              ],
              "parent": undefined,
              "profile": {
                "displayName": "biggroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/id": "eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/realm": "master",
              },
              "name": "subgroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": "biggroup",
              "profile": {
                "displayName": "subgroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/id": "557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/realm": "master",
              },
              "name": "emptygroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": undefined,
              "profile": {
                "displayName": "emptygroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
      ],
      "type": "full",
    },
  ],
]
`;

exports[`KeycloakOrgEntityProvider with v24 should read with grantType client_credential 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/id": "59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/realm": "master",
              },
              "name": "jamesdoe",
            },
            "spec": {
              "memberOf": [
                "biggroup",
              ],
              "profile": {
                "email": "jamesdoe@gmail.com",
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/id": "c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/realm": "master",
              },
              "name": "joedoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/id": "2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/realm": "master",
              },
              "name": "johndoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "displayName": "John Doe",
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/id": "9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/realm": "master",
              },
              "name": "biggroup",
            },
            "spec": {
              "children": [
                "subgroup",
              ],
              "members": [
                "jamesdoe",
              ],
              "parent": undefined,
              "profile": {
                "displayName": "biggroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/id": "eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/realm": "master",
              },
              "name": "subgroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": "biggroup",
              "profile": {
                "displayName": "subgroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/id": "557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/realm": "master",
              },
              "name": "emptygroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": undefined,
              "profile": {
                "displayName": "emptygroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
      ],
      "type": "full",
    },
  ],
]
`;

exports[`KeycloakOrgEntityProvider with v24 should read with grantType password 1`] = `
[
  [
    {
      "entities": [
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/id": "59efec15-a00b-4700-8833-5f4cdecc1132",
                "keycloak.org/realm": "master",
              },
              "name": "jamesdoe",
            },
            "spec": {
              "memberOf": [
                "biggroup",
              ],
              "profile": {
                "email": "jamesdoe@gmail.com",
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/id": "c982b51a-abf6-4f68-bfdf-a1c6257214fc",
                "keycloak.org/realm": "master",
              },
              "name": "joedoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "User",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/users/2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/id": "2bf97dbd-fd6a-47ae-986b-2632fa95e03f",
                "keycloak.org/realm": "master",
              },
              "name": "johndoe",
            },
            "spec": {
              "memberOf": [],
              "profile": {
                "displayName": "John Doe",
                "email": undefined,
              },
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/id": "9cf51b5d-e066-4ed8-940c-dc6da77f81a5",
                "keycloak.org/realm": "master",
              },
              "name": "biggroup",
            },
            "spec": {
              "children": [
                "subgroup",
              ],
              "members": [
                "jamesdoe",
              ],
              "parent": undefined,
              "profile": {
                "displayName": "biggroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/id": "eefa5b46-0509-41d8-b8b3-7ddae9c83632",
                "keycloak.org/realm": "master",
              },
              "name": "subgroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": "biggroup",
              "profile": {
                "displayName": "subgroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
        {
          "entity": {
            "apiVersion": "backstage.io/v1beta1",
            "kind": "Group",
            "metadata": {
              "annotations": {
                "backstage.io/managed-by-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "backstage.io/managed-by-origin-location": "url:http://localhost:8080/admin/realms/master/groups/557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/id": "557501bd-8188-41c0-a2d5-43ff3d5b0258",
                "keycloak.org/realm": "master",
              },
              "name": "emptygroup",
            },
            "spec": {
              "children": [],
              "members": [],
              "parent": undefined,
              "profile": {
                "displayName": "emptygroup",
              },
              "type": "group",
            },
          },
          "locationKey": "keycloak-org-provider:default",
        },
      ],
      "type": "full",
    },
  ],
]
`;
